### PR TITLE
Feat/separate registrations

### PR DESCRIPTION
**Pull request by ZzRanger**
### Overview
Create separate registration for hackers, mentors, and volunteers respectively.

### Notable Changes
After a user signs up for hackportal, they will have the ability to register as a hacker (via /register/hacker), mentor (via /register/mentor), and volunteer (via /register/volunteer). This deviates from the status quo with a single route to register (via /register). After the user finishes registering, their data will be saved to the appropriate firebase collection (hacker, mentor, & volunteer collections respectively).

Mentors & volunteers will be prompted for their availability throughout the hackathon, as seen below.

Screen Shot 2022-06-20 at 8 40 44 AM

### CODE REVIEW

**Comment by abdullah50053** [comment]
Error log:
Screen Shot 2022-06-26 at 12 18 49 AM

**Comment by ZzRanger** [comment]
> Error log: Screen Shot 2022-06-26 at 12 18 49 AM

Looks like this is from the changes to UserAdminView from the most recent commit. My most recent commit should fix the build error.

**Comment by ZzRanger** [comment]
Can I get a review on this PR? I merged #130 onto this branch so once this is done we can close that PR as well.

File: hackportal.config.ts [hunk 1]
```diff
@@ -428,25 +428,47 @@ export const hackPortalConfig: HackPortalConfig = {
...
+
+        textAreaQuestions: [
+          {
+            //don't remove; for user account info
+            question: 'If not, what times are you free?',
+            id: 'timesOptional',
+            name: 'timesOptional',
+            required: false,
+            initialValue: '',
+            placeholder: 'Ex. Saturday 9AM-5PM and Sunday 9AM - 2PM',
```
Comment (DedsecKnight) [inline hackportal.config.ts:463]: Might need to change this to `Date` object later on.

File: lib/types.d.ts [hunk 2]
```diff
@@ -85,30 +85,70 @@ type Registration = {
...
+  // TODO: Allow this to be dynamically defined by the organizers
+  // TODO: responses: { [questionId: string]: Question }
+
   //claims: []; //Array of Strings will be used to id any claims (lunch, merch, etc.) made by user
+} & GeneralInfo &
+  EventInfo;
+
+type AvailabilityInfo = {
+  isFullyAvailabile: boolean;
+  timesAvailabile: string;
```
Comment (DedsecKnight) [inline lib/types.d.ts:139]: Might need to be of another type if we plan to have HackPortal automate shift assignment.

**Review by DedsecKnight — COMMENTED** [review]
Will need a discussion rewriting `scripts/generate-types.ts`, but other than that, LGTM. 

Also, there seems to be a lot of breaking changes in this PR, so not sure if it's better if we merge other stuffs first.

**Review by DedsecKnight — APPROVED** [review]
Looks like there's some merge conflicts, but other than that LGTM!

File: hackportal.config.ts [hunk 3]
```diff
@@ -557,6 +558,13 @@ interface datalistQuestion extends RegistrationQuestion {
   datalist: string;
 }
 
+export type DateRange = { start: Date; end: Date };
+
+interface AvailabilityInfoQuestion extends RegistrationQuestion {
+  value: DateRange[];
+  range: DateRange;
+}
+
```
Comment (DedsecKnight) [inline hackportal.config.ts:567]: Is there any rationale behind having these types in the config file instead of in `@generated/types` ?

**Review by DedsecKnight — DISMISSED** [review]

**Review by DedsecKnight — APPROVED** [review]

**Comment by MichaelZhao21** [comment]
Question: is this necessary?

Basically this just adds a couple of extra questions for mentors/volunteers. I feel like yes it's technically a possible use case but like idk if most other hackathons rly care about using this feature. Like even us we just used our own signupgenius (and we can use some other calendar booking system if we want smth better) but like it works. Plus, I don't think mentors/volunteers rly care about the other stuff and making an account yk. 

Alsoooooooo the merge conflicts 😭 👹 🏃‍♂️

**Comment by abdullah50053** [comment]
Unnecessary

**Pull request closed without merging**